src/main/java/com/commercetools/sync/products/helpers/VariantReferenceResolver.java

```diff
@@ -148,11 +155,17 @@ public VariantReferenceResolver(@Nonnull final ProductSyncOptions productSyncOpt
     private CompletionStage<Void> resolveReference(@Nonnull final JsonNode referenceValue) {
```

Comment:
a generic void type like CompletableFuture<Void> always looks confusing to me, also we have code duplication in here 😄 
Can't we make it better, maybe just return reference and caller could do the then accept part? 🤔

Reply:
> a generic void type like CompletableFuture always looks confusing to me

What is confusing? It is a future that doesn't have result when it finished. Like the difference between executing `Runnable` on an ExecutorService as opposed to `Callable` in old java. 


>Can't we make it better, maybe just return reference and caller could do the then accept part? 

I don't know what you mean. Can you please show example?

Reply:
I don't agree with this: 
> I think it should be enough put void for return type instead of generic CompletableStage

Because there is a big difference between just `void` and `CompletetionStage<Void>` with the later, you know the method contains asynchronous code execution, that you can also block it's execution if you desire. However, with just `void`, it should not contain asyn code, because if it does you won't have control over blocking its execution when needed.

I am not sure I get your reasoning why it's bad design of language. Actually, on the contrast I think it's good design to distinguish between async and sync code. 

> I said it's confusing design because it's calling by `mapValuesToFutureOfCompletedValues`, the second function is a mapper and I would expect the method will return some value instead of mutating the input.

Here, I see your point and how it can be a bit confusing. But I think in this case, one has to accept a mapped value as also being `Void`. Which actually helps us in this case.

> My main question in here we have a code duplication: I just asked can't we make it better without code duplication ?

Yes, I see your point. I committed in another PR which is quite related to this: https://github.com/commercetools/commercetools-sync-java/pull/444/files#diff-925cd7aab9dfa200affec5299f40b548R162-R166

But this other PR is not ready yet for review. Would be ready after this one.

Reply:
ok let me try to explain my point why I don't like the generic void method because it is complex and confusing to understand. Also to make to code async you need to use wrapper classes in java which does not feel good, IMO.  Other languages have some features to write async code easy, for example, c# and js using the `async` keyword instead of using generic wrappers and chain of methods, another example in golang you can put `go` keyword before method call to make it async simple as this. 
But it's understandable that java doesn't want to add these features instead of using weird CompletableFuture API methods, I know java needs to be backward compatible 😄 

For the other PR, it looks good, this is what I was thinking when I reviewing this code.

Reply:
Hmm, I can imagine it could be a bit confusing if you're not used to it. 

But from my perspective, since I am quite used to it. I don't notice how it's confusing at all :D So maybe it's something you just make your head get used to ;)